compat/mingw.c

```diff
@@ -3009,6 +3009,45 @@ static char *wcstoutfdup_startup(char *buffer, const wchar_t *wcs, size_t len)
 	return memcpy(malloc_startup(len), buffer, len);
 }
 
+static void maybe_redirect_std_handle(const wchar_t *key, DWORD std_id, int fd,
+				      DWORD desired_access, DWORD flags)
+{
+	DWORD create_flag = fd ? OPEN_ALWAYS : OPEN_EXISTING;
+	wchar_t buf[MAX_LONG_PATH];
+	DWORD max = ARRAY_SIZE(buf);
+	HANDLE handle;
+	DWORD ret = GetEnvironmentVariableW(key, buf, max);
+
+	if (!ret || ret >= max)
+		return;
+
+	/* make sure this does not leak into child processes */
+	SetEnvironmentVariableW(key, NULL);
+	if (!wcscmp(buf, L"off")) {
+		handle = GetStdHandle(std_id);
+		if (handle != INVALID_HANDLE_VALUE)
+			CloseHandle(handle);
+		return;
+	}
+	handle = CreateFileW(buf, desired_access, 0, NULL, create_flag,
+			flags, NULL);
+	if (handle != INVALID_HANDLE_VALUE) {
+		int new_fd = _open_osfhandle((intptr_t)handle, O_BINARY);
+		SetStdHandle(std_id, handle);
+		dup2(new_fd, fd);
+	}
+}
+
+static void maybe_redirect_std_handles(void)
```

Comment:
I'm not sure about named pipes, they do have slightly different semantics.  But we're caling CreateFileW() here using the pathnames given, so we don't necessarily know they are pipes.

My larger concern is how output is interleaved or overwritten.  For example, if I do:
    git foo >out 2>&1
I expect both streams to get written to "out" and if there is output on both streams, that I'll get all of the output from both -- granted, with a non-deterministic interleaving -- but I'll get all of the content.

If I set both envvars to the same path and do the same command, both will open the same file using 2 completely independent fd's and/or HANDLES -- and therefore have 2 independent seek-pointers.  If both get written to, the beginning of the file will be overwritten by the second writer.
So I'll lose content.  Right?